### ISSUE TITLE
Showing error that comparison is useless, but if changed to non "useless" code it produces different output, meaning its not useless.

### ISSUE THREAD

**Issue by Merlin1846** [rust-lang/rust#94783]
It should not produce a warning but it does.

```rust
let position:(u16,u16) = any_value;
if position.0 >= 0 && position.1 >= 0 {
// do stuff
}
```

The current output is:

```
comparison is useless due to type limits
`#[warn(unused_comparisons)]` on by defaultrustc(unused_comparisons)
```
But it should be:

```
nothing
```

Although this code does not produce a warning it seems a little silly since it does the exact same thing as far as I can tell. But with extra steps.
```rust
let position:(u16,u16) = any_value;
if position.0 as i16 > -1 && position.1 as i16 > -1 {
// do stuff
}
```

And in the event that this is a bug and not just me being stupid should I do this instead?
```rust
let position:(u16,u16) = any_value;
#[allow(unused_comparisons)] {
    if position.0 >= 0 && position.1 >= 0 {
    // do stuff
    }
}
```

**Comment by rtzoeller** [comment]
The warning is valid - `position.0` and `position.1` are `u16`s, i.e. non-negative. Thus `position.0 >= 0 && position.1 >= 0` is always true.

**Comment by Merlin1846** [comment]
The code that does NOT produce a warning produces this result:
![image](https://user-images.githubusercontent.com/57198951/157540975-a6327ef7-46f3-4310-975e-32cf878358b1.png)

But this code, which produces a warning:
```rust
let position:(u16,u16) = any_value;
if position.0 >= 0 && position.1 >= 0 {
// do stuff
}
```
Gives the correct result.
![image](https://user-images.githubusercontent.com/57198951/157541237-dd859474-22a4-463a-a8d9-c78b8b0c0aa5.png)

So that can't be, right?

**Comment by notriddle** [comment]
As proven by [this exhaustive test case](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=d7223e4391e03418e0133703399bfe4c), the following two lines of code mean exactly the same thing:

```rust
let i = a as i16 > -1 && b as i16 > -1;
let u = a < 0x7fff && b < 0x7fff;
```
